File: MAP/Lab 2/ToyInterpreter/src/controller/Controller.java
package controller;

import exceptions.AdtExceptions;
import model.adts.MyIDictionary;
import model.adts.MyIStack;
import model.adts.MyList;
import model.state.PrgState;
import model.statements.IStmt;
import repository.IRepository;
import repository.IRepositoryLog;
import repository.Repository;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class Controller implements IController, IRepository {
    IRepositoryLog repo;

    public Controller(IRepositoryLog repo) {
        this.repo = repo;
    }

    public Controller() {
        repo = new Repository();
    }

    @Override
    public PrgState oneStep(PrgState state) throws Exception {
        MyIStack<IStmt> execStack = state.getExeStack();
        if(execStack.empty())
            throw new Exception("No more operations needed");

        IStmt crtStatement = execStack.pop();
        repo.logPrgStateExec();
        return crtStatement.execute(state);
    }

    @Override
    public PrgState oneStep(int index) throws Exception {
        PrgState state = repo.getProgramAt(index);
        return oneStep(state);
    }

    @Override
    public void allSteps(PrgState state) throws Exception {
        MyIStack<IStmt> execStack = state.getExeStack();
        while(!execStack.empty()) {
            //Execute one step
            oneStep(state);

            //Call Garbage Collector
            state.getHeap().setData(
                    conservativeGarbageCollector(
                            state.getSymTable(),
                            state.getHeap()
                    )
            );
        }
    }

    @Override
    public void allSteps(int index) throws Exception {
        PrgState state = repo.getProgramAt(index);
        allSteps(state);
    }

    @Override
    public String display(PrgState state) {
        return state.toString();
    }

    @Override
    public String display(int index) {
        try {
            return repo.getProgramAt(index).toString();
        } catch (AdtExceptions adtExceptions) {
            adtExceptions.printStackTrace();
        }
        return null;
    }

    @Override
    public HashMap<Integer, Integer> conservativeGarbageCollector(MyIDictionary<String, Integer> symTable, MyIDictionary<Integer, Integer> heap) {
        return (HashMap<Integer, Integer>) heap.entrySet().stream()
                .filter( e -> symTable.getValues().contains(e.getKey()) )
                .collect( Collectors.toMap( Map.Entry::getKey,  Map.Entry::getValue ));
    }

    @Override
    public MyList<PrgState> getPrograms() {
        return repo.getPrograms();
    }

    @Override
    public PrgState getProgramAt(int index) {
        try {
            return repo.getProgramAt(index);
        } catch (AdtExceptions adtExceptions) {
            adtExceptions.printStackTrace();
        }

        return null;
    }

    @Override
    public PrgState newProgram(IStmt stmt) {
        return repo.newProgram(stmt);
    }
}

File: MAP/Lab10/ConsoleApp1/ConsoleApp1/Controller/Controller.cs
using ConsoleApp1.Model.state;
using ConsoleApp1.Model.statements;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using ToyInterpreter;
using ToyInterpreter.Exceptions;
using ToyInterpreter.Model.state;
using ToyInterpreter.Repository;

namespace ToyInterpreter.Controller
{
    class Controller
    {
        private Repository.Repository repo;

        public Controller()
        {
            repo = new Repository.Repository();
        }

        public Controller(Repository.Repository repo)
        {
            this.repo = repo;
        }

        public PrgState oneStep(PrgState state)
        {
            //Get exe stack
            Stack<IStmt> exeStack = state.getExeStack();
            if(exeStack.Count == 0)
            {
                throw new StmtExecException("Exe stack is empty");
            }
                
            //Pop last element from exe stack
            IStmt crtStmt = exeStack.Pop();

            //Execute the element
            return crtStmt.execute(state);

        }

        public void allSteps()
        {
            PrgState crtPrg = repo.getCrtPrg();

            try
            {
                while(true)
                {
                    oneStep(crtPrg);
                    repo.logPrgStateExec(crtPrg);
                }
            }
            catch(StmtExecException e)
            {

            }
        }

        public void addProgram(IStmt prg)
        {
            PrgState newProg = new PrgState(
               new Stack<IStmt>(),
               new Dictionary<String, Int64>(),
               new List<Int64>(),
               prg,
               new Dictionary<Int64, FileTableData>()
             );

            this.repo.addProgram(newProg);
        }

        public PrgState getFirstProgram()
        {
            return this.repo.getCrtPrg();
        }

        public override string ToString()
        {
            return this.repo.getCrtPrg().ToString();
        }
    }
}
